# Predicting Revenue from Parking Citations in Baltimore
Capstone Project for Springboard Data Science Bootcamp

Tamara Monge

### Section 1: Data Wrangling

In [1]:
# import libraries
import pandas as pd
from datetime import datetime, date
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" 

In [2]:
# import datasets: df1, df2
df1 = pd.read_csv('../data/Parking_Citations_2017-09-25.csv')
df2 = pd.read_csv('../data/Parking_Citations_2017-11-30.csv')

/Users/tami/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df1.info()
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1345053 entries, 0 to 1345052
Data columns (total 21 columns):
Citation           1345053 non-null int64
Tag                1344837 non-null object
ExpMM              1300960 non-null object
ExpYY              1345047 non-null float64
State              1345053 non-null object
Make               1343712 non-null object
Address            1345051 non-null object
ViolCode           1344668 non-null float64
Description        1345053 non-null object
ViolFine           1344668 non-null object
ViolDate           1340590 non-null object
Balance            1345053 non-null object
PenaltyDate        0 non-null float64
OpenFine           1344668 non-null object
OpenPenalty        1344668 non-null object
NoticeDate         602001 non-null object
ImportDate         1345053 non-null object
Neighborhood       206670 non-null object
PoliceDistrict     206670 non-null object
CouncilDistrict    206691 non-null float64
Location           1323450 non-nul

As shown above, the two original dataframes were composed of \~2.8 million records and 21 columns: 16 object Series (`Tag`, `ExpMM`, `State`, `Make`, `Address`, `Description`, `ViolFine`, `ViolDate`, `Balance`, `OpenFine`, `OpenPenalty`, `NoticeDate`, `ImportDate`, `Neighborhood`, `PoliceDistrict`, `Location`), 4 float64 Series (`ExpYY`, `ViolCode`, `PenaltyDate`, `CouncilDistrict`) and 1 int64 Series (`Citation`). The original dataframes occupied \~450 MB in size.

In [4]:
# convert ViolDate to datetime object 
df1['ViolDate'] = df1['ViolDate'][pd.notnull(df1['ViolDate'])].apply(lambda x: datetime.strptime(x, '%m/%d/%Y %I:%M:%S %p'))
df2['ViolDate'] = df2['ViolDate'][pd.notnull(df2['ViolDate'])].apply(lambda x: datetime.strptime(x, '%m/%d/%Y %I:%M:%S %p'))

In [5]:
# merge dataframes
df = pd.merge(df1, df2, how='outer')
df = df.drop_duplicates('Citation')
df.shape
df.head()

(1503910, 21)

Citation     Tag ExpMM  ExpYY State   Make                   Address  \
0  79764791  9RX368    00   11.0    WV  INFIN      6000 CHINQUAPIN PKWY   
1  79799128  00T058    06   11.0    MD    GMC  O/S 3600 GLENGYLE AVENUE   
2  79741567  6926L6    03   13.0    MD   FORD              1000 VINE ST   
3  79751210  1FZL87    05   10.0    MD   HOND      4200 CONNECTICUT AVE   
4  79751160  TW2578   SEP   10.0    VA   MERC         4640 COLEHERNE RD   

   ViolCode                         Description ViolFine  \
0      19.0                  Exceeding 48 Hours   $32.00   
1      22.0                        Expired Tags   $32.00   
2      18.0  All Other Parking Meter Violations   $32.00   
3      22.0                        Expired Tags   $32.00   
4      22.0                        Expired Tags   $32.00   

                         ...                          Balance PenaltyDate  \
0                        ...                          $332.00         NaN   
1                        ...                          $332.00         NaN   
2                        ...                          $357.00         NaN   
3                        ...                          $332.00         NaN   
4                        ...                          $332.00         NaN   

   OpenFine OpenPenalty  NoticeDate                    ImportDate  \
0    $32.00     $300.00  10/31/2011  05/09/2013 04:03:00 AM +0000   
1    $32.00     $300.00  01/04/2012  08/09/2013 04:03:00 AM +0000   
2    $32.00     $325.00  10/19/2011  05/09/2013 04:03:00 AM +0000   
3    $32.00     $300.00  08/31/2011  05/07/2013 10:41:00 AM +0000   
4    $32.00     $300.00  12/29/2011  07/09/2013 04:02:00 AM +0000   

  Neighborhood PoliceDistrict CouncilDistrict  \
0          NaN            NaN             NaN   
1          NaN            NaN             NaN   
2          NaN            NaN             NaN   
3    Irvington   Southwestern             8.0   
4          NaN            NaN             NaN   

                                            Location  
0  6000 CHINQUAPIN PKWY\nBaltimore, MD\n(39.36509...  
1  3600 GLENGYLE AVENUE\nBaltimore, MD\n(39.36725...  
2  1000 VINE ST\nBaltimore, MD\n(39.290402, -76.6...  
3  4200 CONNECTICUT AVE\nBaltimore, MD\n(39.28429...  
4  4640 COLEHERNE RD\nBaltimore, MD\n(39.294882, ...  

[5 rows x 21 columns]

In [6]:
# see what df looks like now
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1503910 entries, 0 to 1780483
Data columns (total 21 columns):
Citation           1503910 non-null int64
Tag                1503689 non-null object
ExpMM              1456379 non-null object
ExpYY              1503904 non-null float64
State              1503910 non-null object
Make               1502540 non-null object
Address            1503908 non-null object
ViolCode           1503626 non-null float64
Description        1503910 non-null object
ViolFine           1503626 non-null object
ViolDate           1499548 non-null datetime64[ns]
Balance            1503910 non-null object
PenaltyDate        0 non-null float64
OpenFine           1503626 non-null object
OpenPenalty        1503626 non-null object
NoticeDate         601821 non-null object
ImportDate         1503910 non-null object
Neighborhood       217175 non-null object
PoliceDistrict     217175 non-null object
CouncilDistrict    217196 non-null float64
Location           1482411

The merged dataframe contains \~1.5 million records.

In [7]:
df.columns

Index(['Citation', 'Tag', 'ExpMM', 'ExpYY', 'State', 'Make', 'Address',
       'ViolCode', 'Description', 'ViolFine', 'ViolDate', 'Balance',
       'PenaltyDate', 'OpenFine', 'OpenPenalty', 'NoticeDate', 'ImportDate',
       'Neighborhood', 'PoliceDistrict', 'CouncilDistrict', 'Location'],
      dtype='object')

In [8]:
# drop irrelevant columns
df.drop(['ExpMM', 'ExpYY', 'PenaltyDate', 'NoticeDate', 'ImportDate', 'Neighborhood', 'PoliceDistrict', 'CouncilDistrict'], axis=1, inplace=True)

In [9]:
# rename columns for ease of operations
df.columns =['cit', 'tag', 'state', 'make', 'add', 'code', 'desc', 'fine', 'date', 'bal', 'ofine', 'openalty', 'loc']

In [10]:
# extract relevant cohort
df = df[:][df['date'] >= '2015-09-23']
df.shape

(912308, 13)

Relevant cohort contains 912,308 records.

In [11]:
# break location into longitude and latitude
def find_longitude(location):
    """Ingest location and return longitude"""
    longitude = float('NaN')
    if type(location) == str:
        if ((len(location.split('\n')) > 2) and (location.split('\n')[2] != '')):
            longitude = float(location.split('\n')[2].split('(')[1].split(',')[1].split(')')[0])       
    return longitude
df['lon'] = df['loc'].apply(find_longitude)

def find_latitude(location):
    """Ingest location and return latitude"""
    latitude = float('NaN')
    if type(location) == str:
        if ((len(location.split('\n')) > 2) and (location.split('\n')[2] != '')):
            latitude = float(location.split('\n')[2].split('(')[1].split(',')[0])
    return latitude
df['lat'] = df['loc'].apply(find_latitude)

In [12]:
# build dictionary of address:(longitude,latitude) to fill in rows with missing (lon,lat) data
df['lonlat'] = list(zip(df['lon'], df['lat']))
lonlatadd = list(zip(df['add'], df['lonlat']))
lonlat_dict = dict(lonlatadd)
df['lonlat'] = df['add'].map(lonlat_dict)

Prior to filling in missing lonlat values with dictionary 29.7% were missing; afterwards, 29% were missing; an increase in populated values of 0.7%. This is not a significant increase, so instead we will simply drop the records missing `lonlat.`

In [13]:
# drop records missing lat,lon information
df.dropna(axis=0, how='any', subset=['lat', 'lon'], inplace=True)

In [14]:
df.shape

(641072, 16)

70.3% of records (641,072) contain spatial information.

In [15]:
# define bounding box for baltimore area
northedge = 39.372446
eastedge = -76.529461
southedge = 39.196925
westedge = -76.711344                   

In [16]:
# remove observations outside the baltimore area
df = df[df['lat'] >= southedge]
df = df[df['lat'] <= northedge]
df = df[df['lon'] >= westedge]
df = df[df['lon'] <= eastedge]
df.shape

(623639, 16)

97.3% of records (623,639) fall within Baltimore bounding box.

In [17]:
# break violdate into year, month, day, and hour columns
df['yr'] = df['date'][pd.notnull(df['date'])].dt.year.astype(int)
df['mo'] = df['date'][pd.notnull(df['date'])].dt.month.astype(int)
df['day'] = df['date'][pd.notnull(df['date'])].dt.day.astype(int)
df['hr'] = df['date'][pd.notnull(df['date'])].dt.hour.astype(int)

In [18]:
# clean financial columns
df['fine'] = df['fine'][df['fine'].apply(type) == str].apply(lambda x: x[1:]).astype(float) 
df['bal'] = df['bal'][df['bal'].apply(type) == str].apply(lambda x: x[1:]).astype(float)   
df['ofine'] = df['ofine'][df['ofine'].apply(type) == str].apply(lambda x: x[1:]).astype(float)   
df['openalty'] = df['openalty'][df['openalty'].apply(type) == str].apply(lambda x: x[1:]).astype(float)   

In [19]:
# standardize case and length of string columns
df['state'] = df['state'].str.upper()
df['make'] = df['make'].str.upper()
df['make'] = df['make'][df['make'].apply(type) == str].apply(lambda x: x[:3]) 
df['add'] = df['add'].str.upper()
df['desc'] = df['desc'].str.upper()

Now, let's simplify our `state` and `balance` variables to be binary variables: `instate` and `paid`. 

In [20]:
# simplify balance to a binary variable, paid=[0,1] 
def to_paid(balance):
    """Returns 1 if balance is 0, 0 otherwise."""
    if balance == 0:
        paid = 1
    else:
        paid = 0
    return paid

df['paid'] = df['bal'].apply(to_paid)

In [21]:
# simplify state to a binary variable, instate=[0,1]
def to_instate(state):
    """Returns 1 if state is MD, 0 otherwise."""
    if state == 'MD': 
        instate = 1
    else:
        instate = 0
    return instate

df['instate'] = df['state'].apply(to_instate)

In [22]:
# quadrant classification
def quadrant_classification(lonlat):
    """Ingest latitude, longitude and return city quadrant"""
    citycenter = [-76.6122, 39.2904]   # Center of Baltimore, according to Google
    
    if lonlat[1] >= citycenter[1]:     # NE 
        if lonlat[0] >= citycenter[0]: # NW 
            quadrant = 'NORTHEAST'
        else:
            quadrant = 'NORTHWEST'
    else:
        if lonlat[0] >= citycenter[0]:
            quadrant = 'SOUTHEAST'     # SE  
        else:
            quadrant = 'SOUTHWEST'     # SW
    
    return quadrant

df['quad'] = df['lonlat'].apply(quadrant_classification)

In [23]:
# set DatetimeIndex 
df.index = df['date'] 

In [24]:
# drop missing data based on all remaining columns 
df.dropna(axis=0, how='any', subset=['cit', 'tag', 'state', 'make', 'code', 'desc', 'fine', 'bal', 'ofine', 'openalty', 'loc'], inplace=True)
df.shape

(623208, 23)

In [25]:
# see what df looks like after cleaning
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 623208 entries, 2016-09-17 11:38:00 to 2017-11-07 16:15:00
Data columns (total 23 columns):
cit         623208 non-null int64
tag         623208 non-null object
state       623208 non-null object
make        623208 non-null object
add         623208 non-null object
code        623208 non-null float64
desc        623208 non-null object
fine        623208 non-null float64
date        623208 non-null datetime64[ns]
bal         623208 non-null float64
ofine       623208 non-null float64
openalty    623208 non-null float64
loc         623208 non-null object
lon         623208 non-null float64
lat         623208 non-null float64
lonlat      623208 non-null object
yr          623208 non-null int64
mo          623208 non-null int64
day         623208 non-null int64
hr          623208 non-null int64
paid        623208 non-null int64
instate     623208 non-null int64
quad        623208 non-null object
dtypes: datetime64[ns](1), float64(7), int

In [26]:
df.head(3)

cit      tag state make              add  code  \
date                                                                       
2016-09-17 11:38:00  98076210  U658692    TN  DET  4800 SNADER AVE  99.0   
2016-08-02 13:05:00  97623889  60336CE    MD  ACU    2600 FLEET ST  22.0   
2016-08-02 12:03:00  97623855   9CJM65    MD  TOY     700 BROADWAY  18.0   

                                                         desc  fine  \
date                                                                  
2016-09-17 11:38:00  ALL OTHER STOPPING OR PARKING VIOLATIONS  32.0   
2016-08-02 13:05:00                              EXPIRED TAGS  32.0   
2016-08-02 12:03:00        ALL OTHER PARKING METER VIOLATIONS  32.0   

                                   date    bal    ...            lon  \
date                                              ...                  
2016-09-17 11:38:00 2016-09-17 11:38:00  158.0    ...     -76.677205   
2016-08-02 13:05:00 2016-08-02 13:05:00    0.0    ...     -76.579551   
2016-08-02 12:03:00 2016-08-02 12:03:00    0.0    ...     -76.593496   

                           lat                   lonlat    yr  mo day  hr  \
date                                                                        
2016-09-17 11:38:00  39.340936  (-76.677205, 39.340936)  2016   9  17  11   
2016-08-02 13:05:00  39.285117  (-76.579551, 39.285117)  2016   8   2  13   
2016-08-02 12:03:00  39.283398  (-76.593496, 39.283398)  2016   8   2  12   

                     paid  instate       quad  
date                                           
2016-09-17 11:38:00     0        0  NORTHWEST  
2016-08-02 13:05:00     1        1  SOUTHEAST  
2016-08-02 12:03:00     1        1  SOUTHEAST  

[3 rows x 23 columns]

In [27]:
# save clean data to new file for faster accessibility in the future
filename = 'persistence/cleaned_data_' + str(date.today()) + '.csv'
df.to_csv(filename, date_format='%Y-%m-%d %H:%M:%S')

After cleaning, we have 623,208 complete records (missing data), each with 23 columns.